File: frontend/src/infrastructure/mappers/orderMapper.ts
import Order from "../../domain/models/Order";
import OrderStatus from "../../domain/valueObjects/Order/OrderStatus";
import IOrderApiModel from "../apiModels/IOrderApiModel";
import orderItemMapper from "./orderItemMapper";

const orderMapper = {
    apiToDomain: (source: IOrderApiModel): Order => {
        return new Order({
            id: source.id,
            total: source.total,
            status: OrderStatus.create(source.status),
            dateCreated: new Date(source.dateCreated),
            dateFinished: source.dateFinished == null ? null : new Date(source.dateFinished),
            orderItems: source.orderItems.map(orderItemMapper.apiToDomain),
            serialNumber: source.serialNumber
        });
    },
};

export default orderMapper;

File: Infrastructure/Mappers/OrderMapper.cs
using Domain.Models;
using Domain.ValueObjects.Order;
using Domain.ValueObjects.Shared;
using Infrastructure.DbEntities;

namespace Infrastructure.Mappers;

public static class OrderMapper
{
    public static Order ToDomain(OrderDbEntity source)
    {
        return new Order(
            id: OrderId.ExecuteCreate(source.Id),
            total: Money.ExecuteCreate(source.Total),
            orderSchedule: OrderSchedule.ExecuteCreate(
                dates: OrderDates.ExecuteCreate(dateCreated: source.DateCreated, dateFinished: source.DateFinished),
                status: OrderStatus.ExecuteCreate(source.Status.ToString())
            ),
            orderItems: source.OrderItems.Select(OrderItemMapper.ToDomain).ToList(),
            serialNumber: source.SerialNumber
        );
    }

    public static OrderDbEntity ToDbModel(Order source)
    {
        return new OrderDbEntity(
            id: source.Id.Value,
            total: source.Total.Value,
            dateCreated: source.OrderSchedule.Dates.DateCreated,
            dateFinished: source.OrderSchedule.Dates.DateFinished,
            status: ToDbEntityStatus(source.OrderSchedule.Status),
            serialNumber: source.SerialNumber
        );
    }

    public static OrderDbEntity.Statuses ToDbEntityStatus(OrderStatus status)
    {
        return (OrderDbEntity.Statuses)Enum.Parse(typeof(OrderDbEntity.Statuses), status.Name);
    }
}